extension/src/providers/gotoProvider.ts

```diff
@@ -0,0 +1,67 @@
+import * as vscode from 'vscode';
+import { AutoLispExt } from '../extension';
+import { Sexpression } from '../format/sexpression';
+import { ReadonlyDocument } from '../project/readOnlyDocument';
+
+interface LocatedItem {
+	line: number;
+	column: number;
+	fspath: string;
+}
+
+export class AutolispDefinitionProvider implements vscode.DefinitionProvider{
+	async provideDefinition(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken): Promise<vscode.Location | vscode.Location[]> {
+		const editor: vscode.TextEditor = vscode.window.activeTextEditor;		
+		let selected: string =  editor.document.getText(editor.selection);
+		if (selected === "") {
+			ReadonlyDocument.getMemoryDocument(document).atomsForest.forEach(sexp => {
+				if (sexp instanceof Sexpression && sexp.contains(position)){
+					const atom = sexp.getAtomFromPos(position);
+					if (atom) {
+						selected = atom.symbol;
+					}		
+				}
+			});			
+		}
+		try {
+			return this.findGoToLocations(selected);
+		} catch (error) {
+			return;	// I don't believe this requires a localized error since VSCode has a default "no definition found" response
+		}
+	} 
+
+	// This was designed around the idea of "preference" for opened and workspace documents
+	private findGoToLocations(searchFor: string): vscode.Location[] {
+		const locations: vscode.Location[] = [];
+		let possible = this.findDefunMatches(searchFor, AutoLispExt.Documents.OpenedDocuments.concat(AutoLispExt.Documents.ProjectDocuments));
+		if (possible.length === 0) {
+			possible = this.findDefunMatches(searchFor, AutoLispExt.Documents.WorkspaceDocuments);
+		}
+		possible.forEach(item => {			
+			locations.push(new vscode.Location(vscode.Uri.file(item.fspath), new vscode.Position(item.line, item.column)));
+		});	
+		return locations;
+	}
+
+	private findDefunMatches(searchFor: string, searchIn: ReadonlyDocument[]): LocatedItem[] {
+		const result: LocatedItem[] = [];
+		searchIn.forEach((doc: ReadonlyDocument) => {
+			doc.atomsForest.forEach(atom => {
+				if (atom instanceof Sexpression){
+					const defs = atom.findChildren(/^DEFUN$|^DEFUN-Q$/i, true);
```

Comment:
It looks like we only take care the function. Maybe to take care the global (`setq`) or function local (setq) is also needed.

```
(defun sampleFunc (x y / a b c d)
  (setq a (list 1 2 3 4) ; does this screw it all up?
        d (setq b 0)
        b (mapcar '+ a))
  (foreach x a
    (setq d (1+ d)) ;; select d and hit F12, maybe it should goto the first line
  )
```

Reply:
I was thinking that too, but then I got concerned the implications of doing that could easily block simpler function names from ever being located in the workspace. I guess if I check to see their Sexpression.Atoms[1].symbol is the match it could prefer functions and if that is false it can prefer variables headers/declarations.